### ISSUE TITLE
Date parsing timezone issues lead to user confusion

### ISSUE THREAD

**Issue by ervwalter** [jeremyfa/yaml.js#15]
This issue relates to a problem being experienced by some DocPad users (see bevry/docpad#321).

yamljs parses date-like properties and converts them to javascript Date objects using `new Date(string)`:

https://github.com/jeremyfa/yaml.js/blob/master/src/yaml/YamlInline.js#L429-430

This is a useful behavior, but it is affected by the underlying inconsistent behavior of Javascript itself.  Concider these three properties:

```
date: 2013-10-14
dateTime: 2013-10-14 00:00:00
dateTime2: 2013-10-13 19:00:00
```

The problem with `new Date(string)` is that it is inconsistent in what timezone it assumes.  For strings that are just a date (e.g. `date`), it returns a date that represents midnight _UTC_.  For strings that include a date and a time (e.g. `dateTime` and `dateTime2`), it returns a date that represents the specified time in the _local_ timezone.

This behavior is odd, but it is what it is.  It becomes troublesome in the context of yaml.js because it is no longer possible to tell from the resulting JSON what was intended by the author of the YAML.

Concider the metadata above run through yamls.js on a browser running in the central timezone (US).  You get these results:

```
date: Sun Oct 13 2013 19:00:00 GMT-0500 (Central Daylight Time)
dateTime: Mon Oct 14 2013 00:00:00 GMT-0500 (Central Daylight Time)
dateTime2: Sun Oct 13 2013 19:00:00 GMT-0500 (Central Daylight Time)
```

The challenge is that you can not distinguish between `date` and `dateTime2` and it's highly likely that the author of the YAML did not intend them to be the same date.  I would suggest that most authors probably would have expected that `date` and `dateTime` were the two that would end up being the same value.  

As you can see in bevry/docpad#321, some DocPad users have bumped into this.  The DocPad code can't workaround it at the moment, because by the time the DocPad code sees the parsed YAML, it is not possible to tell if the author typed just a lone date.

Are you open to changes to yamljs to address this?  I can think of a couple possible options:
- Change yaml.js to explicitly handle date strings that don't have times and force parse them as midnight local time.  Since dates with times are assumed to be local times, this would make things consistent.  This is what [moment.js](http://momentjs.com/docs/#/parsing/string/) does, for example.  If backwards compatibility is a concern, this new behavior could be optional based on some flag.
- Add an option to yaml.js to disable automatic date parsing so that DocPad gets access to the original string value and can handle converting it to a Date object itself.

Thoughts?

**Comment by jeremyfa** [comment]
I am currently working on a major rewrite of the parser that I will release in less than a month.
The rewrite will include a fix for this issue.
As you suggested, it will use midnight local time when no explicit time zone is given. This will ensure consistency.

**Comment by arcanis** [comment]
@jeremyfa Unrelated to this issue, but will this rewrite include some kind of custom type support ?

**Comment by jeremyfa** [comment]
@arcanis Yes!

**Comment by turizoft** [comment]
I support this. I looked for a long while why moment.js `isValid` method didn't return true on the generated JSON date variables. Then I figured out that yaml.js was parsing dates by itself. An option to avoid parsing dates or to at least specify the format would be nice.

**Comment by jeremyfa** [comment]
The newly released version `v0.2.0` should fix it.
